icu4c/source/i18n/gregoimp.cpp

```diff
@@ -33,28 +33,33 @@ int64_t ClockMath::floorDivide(int64_t numerator, int64_t denominator) {
 }
 
 int32_t ClockMath::floorDivide(double numerator, int32_t denominator,
-                          int32_t& remainder) {
-    double quotient;
-    quotient = uprv_floor(numerator / denominator);
-    remainder = (int32_t) (numerator - (quotient * denominator));
+                          int32_t* remainder) {
```

Comment:
FYI
I see that some of these output parameters got changed from references to pointers, even though they are still required.

We should use references when they are required, and pointers when they are optional (can be nullptr), as we have always done.

Reply:
I think what we have always done is very bad due to the fact that by reading the code it is very difficult to realize the call will have side effect to change the value of that parameter. I understand that is what we always did but I think that is a good receipe for buggy code. The worst case for passing as pointer is the caller pass in nullptr and the code just crash right away, but the worst case for passing in as reference is the reader cannot tell that function call will change the value easily unless the reader have to read all the header for every single function call in the code.

Reply:
> I think what we have always done is very bad due to the fact that by reading the code it is very difficult to realize the call will have side effect to change the value of that parameter. I understand that is what we always did but I think that is a good receipe for buggy code. The worst case for passing as pointer is the caller pass in nullptr and the code just crash right away, but the worst case for passing in as reference is the reader cannot tell that function call will change the value easily unless the reader have to read all the header for every single function call in the code.

The rule that we have followed is that if it's nullable, use a pointer; otherwise, use a reference.
We have not really had any problems with this convention. I would rather not churn the code in this way.

Reply:
I'm going to chime in here to side with Markus.  I agree with Frank that the current reference syntax is not self-documenting and can lead to errors, but pointers in general are a pain in the butt and a source of errors, and consistency across the library is vital.

Reply:
> We have not really had any problems with this convention. I would rather not churn the code in this way.

The problem happen when people read the code and debugging the code . The fact that it fool me while I debug this problem and cost me more time to address this issue itself is the problem. I would spot the issue much earlier if the function call is not an output parameter passing by reference.

Reply:
I file https://unicode-org.atlassian.net/browse/ICU-22058 . Let's follow up there.